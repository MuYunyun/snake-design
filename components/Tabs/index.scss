@import '../styles/themes/default';

$tabs: 'tabs';

.#{$tabs} {
  box-sizing: border-box;

  &-wrap {
    display: flex;
    position: relative;
    background: #fff;
    overflow-x: scroll;

    &::-webkit-scrollbar {
      display: none; // 移除滚动条
    }

    &-top {
      height: 40px;
    }

    &-bottom {
      height: 40px;
    }

    &-left {
      flex-direction: column;
      width: 60px;
    }

    &-right {
      flex-direction: column;
      width: 60px;
    }

    &-item {
      flex: 1;
      position: relative;
      display: flex;
      justify-content: center;
      align-items: center;
      color: #666;

      &-customWidth {
        flex: initial;
      }

      &-top {
        &::after {
          position: absolute;
          content: '';
          background-color: #ddd;
          left: 0;
          bottom: 0;
          width: 100%;
          height: 1px;
          transform: scaleY(0.5);
        }
      }

      &-bottom {
        &::after {
          position: absolute;
          content: '';
          background-color: #ddd;
          left: 0;
          top: 0;
          width: 100%;
          height: 1px;
          transform: scaleY(0.5);
        }
      }

      &-left {
        &::after {
          position: absolute;
          content: '';
          background-color: #ddd;
          right: 0;
          height: 100%;
          width: 1px;
          transform: scaleX(0.5);
        }
      }

      &-right {
        &::after {
          position: absolute;
          content: '';
          background-color: #ddd;
          left: 0;
          height: 100%;
          width: 1px;
          transform: scaleX(0.5);
        }
      }
    }

    &-scroll {
      flex-shrink: 0;
    }

    &-active {
      color: #333;
    }

    &-underline {
      position: absolute;
      border: 1px solid #333;
      z-index: 10;

      &-top {
        bottom: 0;
        transition: left 0.3s cubic-bezier(0.35, 0, 0.25, 1), width 0.3s cubic-bezier(0.35, 0, 0.25, 1);
      }

      &-right {
        left: 0;
        transition: top 0.3s cubic-bezier(0.35, 0, 0.25, 1), width 0.3s cubic-bezier(0.35, 0, 0.25, 1);
      }

      &-bottom {
        top: 0;
        transition: left 0.3s cubic-bezier(0.35, 0, 0.25, 1), width 0.3s cubic-bezier(0.35, 0, 0.25, 1);
      }

      &-left {
        right: 0;
        border: 1px solid $color-brand-normal;
        transition: top 0.3s cubic-bezier(0.35, 0, 0.25, 1), width 0.3s cubic-bezier(0.35, 0, 0.25, 1);
      }
    }
  }

  &-left,
  &-right {
    display: flex;
  }
}
