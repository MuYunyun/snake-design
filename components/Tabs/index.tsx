import * as React from 'react'
// import { TabsProps } from 'types/tabs.d'
import ClassNames from 'classnames'

const { useEffect } = React

function Tabs(props: any) {
  const {
    activeTab,
    tabs,
    tabBarPosition,
    tabBarActiveTextColor,
    tabBarInactiveTextColor,
    notEqualDvided,
    className,
    onChange,
    onTabClick
  } = props

  const prefixCls = 'tabs'
  // let leftDistance = 0
  // let tab0

  useEffect(() => {
    // leftDistance = (this as any)[`tab${0}`] && (this as any)[`tab${0}`].getBoundingClientRect().left
    // leftDistance = 0
    handleUnderline()
  }, [activeTab])

  // 处理下划线样式
  function handleUnderline() {
    if (tabBarPosition === 'top' || tabBarPosition === 'bottom') {
      ;(this as any).underline &&
        ((this as any).underline.style.width = `${getUnderlineWidth(activeTab)}px`)
      ;(this as any).underline && ((this as any).underline.style.left = `${getLeft(activeTab)}px`)
    }
  }

  // 获取下划线宽度
  function getUnderlineWidth(_current: number) {
    // let underlineWidth = (this as any)[`tab${current}`].getBoundingClientRect().width
    let underlineWidth = 0
    return underlineWidth
  }

  // 获取当前激活 tab 距离左侧的距离
  function getLeft(_current: number) {
    // return current * (leftDistance * 2 + (this as any)[`tab${current}`].getBoundingClientRect().width) + leftDistance
    return 0
  }

  // 更改 Tab
  function changeTab(e: React.MouseEvent<HTMLElement>, index: number, r: any) {
    // 禁用 tab 点击
    if (r && r.disable) return
    // 下划线滑动逻辑
    underlineMove(index)

    if (index !== activeTab) {
      onChange && onChange(index, e)
    }
    onTabClick && onTabClick(index, e)
  }

  // 下划线滑动逻辑
  function underlineMove(_index: number) {}

  function renderTab() {
    return (
      <div className={ClassNames(`${prefixCls}-wrap`, `${prefixCls}-wrap-${tabBarPosition}`)}>
        {tabs.map((r: any, index: number) => {
          return (
            <div
              className={ClassNames(
                `${prefixCls}-wrap-item`,
                `${prefixCls}-wrap-item-${tabBarPosition}`,
                {
                  [`${prefixCls}-wrap-item-customWidth`]: notEqualDvided,
                  [`${prefixCls}-wrap-scroll`]: tabs.length > 4 // 大于 4 则开启滑动
                }
              )}
              onClick={e => changeTab(e, index, r)}
              key={index}
              style={{ flexBasis: `${(1 / 4) * 100}%` }}
            >
              <div
                className={ClassNames({
                  [`${prefixCls}-wrap-active`]: activeTab === index
                })}
                style={
                  activeTab === index
                    ? { color: tabBarActiveTextColor }
                    : { color: tabBarInactiveTextColor }
                }
                // ref={(ele: any) => {`tab${index}` = ele }}
                // ref={(ele: any) => { tab0 = ele }} // todo 如何动态声明出 tab0 ~ tab10
              >
                {r.title || ''}
              </div>
            </div>
          )
        })}
        <div
          className={ClassNames(
            `${prefixCls}-wrap-underline`,
            `${prefixCls}-wrap-underline-${tabBarPosition}`
          )}
          // ref={ele => { (this as any).underline = ele }}
        />
      </div>
    )
  }

  function renderContent() {
    return <div>This is a swipe</div>
  }

  function layout() {
    switch (tabBarPosition) {
      case 'left':
        return (
          <div className={`${prefixCls}-left`}>
            {renderTab()}
            {renderContent()}
          </div>
        )
      case 'right':
        return (
          <div className={`${prefixCls}-right`}>
            {renderContent()}
            {renderTab()}
          </div>
        )
      case 'bottom':
        return (
          <div>
            {renderContent()}
            {renderTab()}
          </div>
        )
      default:
        return (
          <div>
            {renderTab()}
            {renderContent()}
          </div>
        )
    }
  }

  function getClassName() {
    return ClassNames(`${prefixCls}`, className)
  }

  return <div className={getClassName()}>{layout()}</div>
}

export default Tabs
