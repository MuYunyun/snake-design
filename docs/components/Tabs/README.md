---
imports:
  import TabsDemoCode from '!raw-loader!./simpleTabs.tsx';
  import TabDemo from './simpleTabs.tsx';
  import Code from '../../layout/prism/index'
---

# Tabs 标签页

选项卡切换组件。

### 简单示例

:::demo
各种类型的按钮

<TabDemo />
:::

## 代码示例

<div><Code>{TabsDemoCode}</Code></div>

## Props
| 参数     | 说明              | 类型   | 可选值 | 默认值 |
| -------- | ----------------- | ------ | ------ | ------ |
| className | `按钮类名` | string | - | - |
