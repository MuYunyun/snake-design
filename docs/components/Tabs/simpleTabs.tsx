import * as React from 'react'
import Tabs from 'components/Tabs'
import './demo.scss'

const tabs0 = [
  { title: '标签一' },
  { title: '标签二' },
  { title: '标签三' },
];

export default class SimpleTab extends React.Component {

  state = {
    page0: 0
  }

  onChange = (index: number, e: any, order: any) => {
    this.setState({
      [`page${order}`]: index
    })
  }

  render() {
    const { page0 } = this.state

    return (
      <>
        <Tabs
          tabs={tabs0}
          activeTab={page0}
          onChange={(index: number, e: any) => this.onChange(index, e, 0)}
          style={{width: 500}}
        >
          <div style={{ display: 'flex', alignItems: 'center', justifyContent: 'center', height: '120px', backgroundColor: '#fff' }}>
            标签一的内容
          </div>
          <div style={{ display: 'flex', alignItems: 'center', justifyContent: 'center', height: '120px', backgroundColor: '#fff' }}>
            标签二的内容
          </div>
          <div style={{ display: 'flex', alignItems: 'center', justifyContent: 'center', height: '120px', backgroundColor: '#fff' }}>
            标签三的内容
          </div>
        </Tabs>
      </>
    )
  }
}
