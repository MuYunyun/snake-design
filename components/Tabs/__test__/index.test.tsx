import * as React from 'react'
import { render, fireEvent } from 'react-testing-library'
import Tabs from '../index'

const { useState } = React

const tabs = [{ title: '标签一' }, { title: '标签二' }, { title: '标签三', disabled: true }]

describe('Tabs Test', () => {
  beforeEach(() => {
    document.body.innerHTML = ''
  })
  it('snapshot', () => {
    const tabsRenderTop = render(
      <Tabs tabs={tabs}>
        <div>标签一的内容</div>
        <div>标签二的内容</div>
        <div>标签三的内容</div>
      </Tabs>
    )
    const tabsRenderBottom = render(
      <Tabs tabs={tabs} tabBarPosition="bottom">
        <div>标签一的内容</div>
        <div>标签二的内容</div>
        <div>标签三的内容</div>
      </Tabs>
    )
    const tabsRenderLeft = render(
      <Tabs tabs={tabs} tabBarPosition="left">
        <div>标签一的内容</div>
        <div>标签二的内容</div>
        <div>标签三的内容</div>
      </Tabs>
    )
    const tabsRenderRight = render(
      <Tabs tabs={tabs} tabBarPosition="right">
        <div>标签一的内容</div>
        <div>标签二的内容</div>
        <div>标签三的内容</div>
      </Tabs>
    )
    expect(tabsRenderTop.container.innerHTML).toMatchSnapshot()
    expect(tabsRenderBottom.container.innerHTML).toMatchSnapshot()
    expect(tabsRenderLeft.container.innerHTML).toMatchSnapshot()
    expect(tabsRenderRight.container.innerHTML).toMatchSnapshot()
  })
  it('dispatch change tab', async () => {
    const TabDemo = function() {
      const [activeTab, setActiveTab] = useState(0)
      return (
        <Tabs tabs={tabs} activeTab={activeTab} onChange={(index: number) => setActiveTab(index)}>
          <div>标签一的内容</div>
          <div>标签二的内容</div>
          <div>标签三的内容</div>
        </Tabs>
      )
    }

    const { getByText } = render(<TabDemo />)
    expect(getByText('标签一的内容')).toBeTruthy()
    fireEvent.click(getByText('标签二'))
    expect(getByText('标签二的内容')).toBeTruthy()
    fireEvent.click(getByText('标签三'))
    expect(getByText('标签二的内容')).toBeTruthy()
  })
})
