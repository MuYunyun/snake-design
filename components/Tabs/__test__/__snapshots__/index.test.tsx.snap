// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component renders a primary button correctly 1`] = `
<div>
  button
</div>
`;
